python/ray/data/dataset.py

```diff
@@ -462,74 +463,34 @@ def repartition(self, num_blocks: int, *, shuffle: bool = False) -> "Dataset[T]"
 
         if shuffle:
 
-            def do_shuffle(blocks, clear_input_blocks: bool):
-                # TODO: implement clear_input_blocks
-                return simple_shuffle(blocks, num_blocks)
+            def do_shuffle(block_list, clear_input_blocks: bool, block_udf):
```

Comment:
Add the concept of a block_udf which is later used to implement fusion of normal and map shuffle stages.